### ISSUE TITLE
Running in a JVM application with LD_PRELOAD

### ISSUE THREAD

**Issue by rodesai** [jemalloc/jemalloc#1919]
I want to replace glibc's malloc with jemalloc in my Java application to fix excessive RSS we've seen with glibc. I'm thinking that I can just do this by setting LD_PRELOAD to the path to jemalloc, and empirically this seems to work just fine. I was wondering if this is a common/recommended pattern, and whether there are any gotchas I should watch out for.

Fora bit more context, what's making me a bit nervous is that the docs for tcmalloc mention that replacing the allocator this way is not something they necessarily recommend, but don't really go into detail why. AFAICT there are a few possible issues:
- possibly using 2 allocators in the same process, which might conflict on sbrk usage, or mix up buffers when calling `free`. I've confirmed that all the binaries in the running app have a runtime dependency on glibc so I'm not too worried here. I'm also running with sbrk disabled just to be safe.
- LD_PRELOAD is ignored when setuid is set. I don't care about this.
- LD_PRELOAD is inherited by forked processes. We don't fork so I'm not worried about this.

tl;dr - is running a java application w/ jemalloc using LD_PRELOAD a common pattern? any other gotchas I should be aware of?

**Comment by davidtgoldblatt** [comment]
It's doable, and we have some experience with it, but there are some risks; we've uncovered a couple of Hotspot bugs that only repro with jemalloc:
- https://github.com/jemalloc/jemalloc/issues/1392 and its corresponding upstream bug (this is fixed in the upstream JVM, but only in recent releases).
- A concurrent fork bug in java.lang.ProcessImpl.forkAndExec (AFAIK this isn't actually jemalloc specific, it just seems to repro more frequently there). I'm not actually sure if this has been reported upstream.

**Comment by rodesai** [comment]
Thanks, this is helpful! I'll make sure we run at a bugfix that fixes JDK-8215355.

**Comment by interwq** [comment]
Closing but feel free to ask if anything.

**Comment by YSavitski** [comment]
> I want to replace glibc's malloc with jemalloc in my Java application to fix excessive RSS we've seen with glibc. I'm thinking that I can just do this by setting LD_PRELOAD to the path to jemalloc, and empirically this seems to work just fine. I was wondering if this is a common/recommended pattern, and whether there are any gotchas I should watch out for.
> 
> Fora bit more context, what's making me a bit nervous is that the docs for tcmalloc mention that replacing the allocator this way is not something they necessarily recommend, but don't really go into detail why. AFAICT there are a few possible issues:
> 
> * possibly using 2 allocators in the same process, which might conflict on sbrk usage, or mix up buffers when calling `free`. I've confirmed that all the binaries in the running app have a runtime dependency on glibc so I'm not too worried here. I'm also running with sbrk disabled just to be safe.
> * LD_PRELOAD is ignored when setuid is set. I don't care about this.
> * LD_PRELOAD is inherited by forked processes. We don't fork so I'm not worried about this.
> 
> tl;dr - is running a java application w/ jemalloc using LD_PRELOAD a common pattern? any other gotchas I should be aware of?

Hi @rodesai , could you please explain how do you execute your java app using jemalloc and LD_PRELOAD? I Tried to add LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.1 as VM argument but got an exception like  _Could not find or load main class LD_PRELOAD=.usr.lib.x86_64-linux-gnu.libjemalloc.so_
